### ISSUE TITLE
B2 efficiency

### ISSUE THREAD

**Issue by fracai** [gilbertchen/duplicacy#179]
Let me preface this by saying I hope I got all my math right. Corrections are welcome. I'm attaching the python script I used for calculations and a table of those results.

tl;dr: **b2_list_file_names** is inefficient to check for existing chunks. Instead use **b2_download_file_by_name** with a header "Range: bytes=0-0".

Duplicacy calls **b2_list_file_names** prior to every upload to check if a given chunk has already been uploaded. As referenced in #30, the B2 pricing means around 10 GB can be uploaded for free each day, and then $.001/GB. However, the list query only returns up to 1000 entries each call so multiple calls per chunk must be made once more than 1000 chunks have been uploaded. 1000 chunks works out to around only 4 GB. By the time I reach 1 TB of data (around 263000 chunks) it's going to be 263 calls to **b2_list_file_names** for each new chunk. Ignoring the free data and transactions (negligible given the total data) it's going to cost almost $140 to make the 35 million calls while uploading. 10 TB would require 2.5 million chunks, 3.5 billion total calls, cost $13,750, and require another 2622 list calls just to upload another new chunk. That's over a penny every 4 MB.

Now, there is the **b2_get_file_info** call, but that requires passing the **fileId** which is normally obtained from **b2_list_file_names**. Duplicacy could save the **fileId** in the snapshots files, but that would require an architecture change.

Alternatively, Duplicacy could call **b2_download_file_by_name** and look at the response code (200 vs 400) and headers. A GET or HEAD adds the entire payload size to your download quota, but including a Range header only adds the requested Range size. Setting a "Range: bytes=0-0" is only a single byte transferred. This reduces the cost by an order of magnitude and also is guaranteed to require a single call, rather than multiple calls. This reduces the cost to upload 10 TB from $13,750 to $1.05. The additional single byte (GET or HEAD) equals around 2 billion calls for a penny. Conveniently, the response headers include the relevant info also returned by **b2_get_file_info**. It looks like the only real difference is a lack of the bucket id.

On the more hacky side, **b2_hide_file** and **b2_delete_file_version** are each unbilled calls. These could be combined to test for the existence of a file. I can see an issue if the process dies before unhiding the chunk and it looks like Duplicacy uses B2 hidden files for fossils. I think this option is more trouble than it's worth.

More generally, B2 chunks are all stored in the "chunks" prefix. If these were instead broken up by nested prefixes taken from the hash, the **b2_list_file_names** could be made more stable by using the relevant prefix when calling. In other words, _chunk/badcoffee12345_ would instead be stored at _chunk/ba/dc/of/fe/e1/23/45_. Each prefix, and therefore each call to **b2_list_file_names**, would hold no more than 256 files. 10 TB would now cost $10.49. Switching to **b2_download_by_name** would be a tenth further, but wouldn't benefit any further than using the download call with all chunks in the same directory.

Changing the file structure would break old backups, but perhaps it could be implemented as new storage target.

I also sent a message to Backblaze to ask about new API calls that could be used to check for file existence. Specifically:
 - **b2_get_file_info_by_name** 
 - **b2_get_file_id_by_name**
I received a response that it's been added as a feature request.

[b2-calc.txt](https://github.com/gilbertchen/duplicacy/files/1290203/b2-calc.txt)
[b2-calc.py.txt](https://github.com/gilbertchen/duplicacy/files/1290204/b2-calc.py.txt)

**Comment by jkl1337** [comment]
While the use of `b2_download_by_name` is interesting, what you overlooked is that `b2_list_file_names` takes a prefix argument (`startFileName`), and this prefix doesn't have to be limited by folder structure. You will see in `duplicacy_b2client.go` code that for probing for a single file, the prefix is set to the file name, so only 1 or 0 results should be returned from each call in this case. In the end, the number of calls to `b2_list_file_names` will be the number of chunks uploaded.

You had me panicked for a bit!

**Comment by fracai** [comment]
Hmmm, that would indeed significantly reduce the required calls, but not
exactyl how you've suggested. **b2_list_file_names** does take a prefix
argument, but that's not what startFileName is. **startFileName** is the
first file that should be returned in the list. It's a way to incrementally
step through the list of all files if there are more than 1000 in B2. The
**b2_list_file_names** response includes a **nextFileName** paramater that
can be used to request the next set of names. The prefix argument may
indeed allow passing in just the chunk name and path. I've been writing
python tests to play with the B2 API and I'll try this this evening.

Regardless, it does indeed look like the call to ListFileNames uses the
startFileName to only make a single call. There is a **singleFile** boolean
that is passed in as true from the GetFileInfo method.

Thanks for the correction, things are indeed not nearly as worrisome as I
thought. Using **b2_download_by_name** could still be used to cut upload
costs to a tenth of what they are now, but it's indeed not nearly as
pressing of an issue. I'll update my calculator and post that this evening
as well.

Very sorry for the alarm.

**Comment by jkl1337** [comment]
startFileName is the first file that should be returned... But it is the first name. So if you give it the exact name of an existing file, that will be the first name returned. If the file doesn't exist, you will get something else. Please look at the docs.. there is no stipulation that startFileName be used only for iteration. It can be used to look up by any prefix quickly. It is basically the interface to a trie.

**Comment by fracai** [comment]
Agreed, I thought I said as much in my response. Sorry if that wasn't clear
previously. I'm not concerned about excess calls anymore, though using
b2_download_by_name would still reduce costs.

**Comment by jkl1337** [comment]
Yeah, and sorry I was conflating the `prefix` and `startFileName` arguments! I was looking at the docs for prefix.. Then the source of duplicacy uses `startFileName`. Which is of course enough for exact name lookup, but there is a separate arg for prefix!

**Comment by jkl1337** [comment]
The 0 byte range download does seem like a nice idea if request latency is not too bad (or better).

**Comment by fracai** [comment]
I ran an informal test where I made 100 calls to each of GET
b2_list_file_names, GET b2_download_by_name, and HEAD b2_download_by_name.
They were all close enough in time that I think it's safe to say any
difference would be negligible.

**Comment by gilbertchen** [comment]
Great discussion.  I agree b2_download_by_name is a great idea and will give it a try.

**Comment by fracai** [comment]
I look forward to the results!

I updated my test script and verified **b2_list_file_names** does indeed accept a full chunk pathname as the **prefix** and then only returns that file information. If **b2_download_by_name** doesn't pan out, that might be a simplification to the GetFileInfo call in terms of processing the response.

And, attached is the updated calculation script and output. It shows that it's about $1.05/TB to upload (not bad at all).
[b2-calc.py.txt](https://github.com/gilbertchen/duplicacy/files/1294495/b2-calc.py.txt)
[b2-calc.txt](https://github.com/gilbertchen/duplicacy/files/1294496/b2-calc.txt)

**Comment by flamingm0e** [comment]
@fracai, it's good to see you here!  I look forward to seeing what comes of Duplicacy.  It has great potential to fill a gap left by the Crashplan debacle

**Comment by fracai** [comment]
@flamingm0e I'm trying, but failing, to place your name or avatar. Do I
know you from somewhere? (feel free to fork this to email "@gmail.com")
Crashplan is definitely why I'm here as well. I'm preferring Duplicacy
already.

**Comment by fracai** [comment]
Using the download call has been merged, but a release hasn't been posted yet. You could indeed reduce the backup frequency or compile your own binary from source.

There have been a lot of changes to the source since the last release so I would expect a new release soon, but I think @gilbertchen is first working on vendoring the repository.

**Comment by Fonant** [comment]
Compiling my own binary for the latest source worked well :simple_smile: Now uses Class B instead of Class C transactions for a ten-fold reduction in transaction costs.

**Comment by compeek** [comment]
Should this issue be closed? I found it wondering about B2 efficiency as well, but it sounds like this was already fixed in a recent version.

**Comment by TheBestPessimist** [comment]
@gilbertchen +1 for closing this.

**Comment by sshaikh** [comment]
I came here to investigate the large number of Class B calls Duplicacy makes while backing up - I'm coming from a place of ignorance, but I was still curious as to why so many calls are needed at all.

Can't a single cache or index file be used instead? There'd need to be some kind of integrity check or process but I suspect that it could be done. If this can't be done because of the lock free features, perhaps an option to put it in "single client" mode for those of us in that position?

I know we're talking about pennies, but it really does seem a bit unfortunate to have to spend on transactions even if the data set hasn't changed at all.

**Comment by gilbertchen** [comment]
Only the initial backup will try to list all chunks existing in the storage and this will generate a number of Class B calls depending on the number of chunks.  So if the initial backup keeps failing then it will try to list the chunks every time.

Once the initial backup is done, the chunks from the last backup will be used as the cache, and no listing of the storage will done.  In the current implementation, only Class C API is used to check if a chunk to be uploaded exists in the storage.  Such chunks will be only issued to new chunks (those not in the last backup).

If you have more questions please create a new issue.
